### PR TITLE
Clessig/develop/channel logging 282

### LINKED ISSUE #282: Implement logging loss per-chanel
### Is your feature request related to a problem? Please describe.

Currently there's no way to monitor how individual chanels and loss functions are changing during training. This is critical for debugging multi-stream models during training (e.g., isolating unstable channels, finiding channels that perform worse). We can make this optional via a flag to reduce clutter. This was part of issue https://github.com/ecmwf/WeatherGenerator/issues/29

### Describe the solution you'd like

Loss is computed individually [here](https://github.com/ecmwf/WeatherGenerator/blob/65f9c85b7f34f1ac7adcc44d467e8bd3f4377442/src/weathergen/train/trainer.py#L437C49-L437C68) and averaged over channels to be stored in [`losses_all`](https://github.com/ecmwf/WeatherGenerator/blob/65f9c85b7f34f1ac7adcc44d467e8bd3f4377442/src/weathergen/train/trainer.py#L442C25-L442C54). 

The solution is to add another dimension to `losses_all` and save the individual losses there. As this array is used anyway for logging, minimal changes is required within the code. This change would fully backwards compatible.

### Additional context

_No response_

### Organisation

AWI

### PR DESCRIPTION
**Pull request by clessig**
## Description

This is @kacpnowak PR (#440) with some minor fixes

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [X] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Closes #282

## Code Compatibility

-   [X] I have performed a self-review of my code

### Code Performance and Testing

-   [X] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [ ] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

File: config/streams/streams_ocean/fesom.yml
Comment (MatKbauer) [inline config/streams/streams_ocean/fesom.yml:1]: Is `test4.zarr` here just for testing/debugging purposes, or should it stay here?

File: src/weathergen/utils/distributed.py [hunk 1]
```diff
@@ -33,3 +34,86 @@ def is_root(pg: dist.ProcessGroup | None = None) -> bool:
...
+    Average a tensor across DDP ranks
+
+    Params:
+        data: tensor to be averaged (arbitrary shape)
+
+    Return :
+        tensor with same shape as data, but entries averaged across all DDP ranks
+    """
+    if _is_distributed_initialized():
+        dist.all_reduce(data.cuda(), op=torch.distributed.ReduceOp.AVG)
```
Comment (MatKbauer) [inline src/weathergen/utils/distributed.py]: let's use the imported shorthand for `torch.distributed` here, that is, `dist.ReduceOp.AVG`

File: src/weathergen/utils/distributed.py
Comment (MatKbauer) [inline src/weathergen/utils/distributed.py:1]: No strong opinion, but we could `from torch import Tensor` and replace `torch.Tensor` by `Tensor`. Have seen this in other places of the code.

File: src/weathergen/utils/distributed.py [hunk 2]
```diff
@@ -33,3 +34,86 @@ def is_root(pg: dist.ProcessGroup | None = None) -> bool:
...
+    """
+    Get MPI world size
+
+    Returns:
+        int: world size
+    """
+    if not dist.is_available():
+        return 1
+    if not dist.is_initialized():
+        return 1
```
Comment (MatKbauer) [inline src/weathergen/utils/distributed.py:47]: Cen we use `if not (dist.is_available() or dist.is_initialized()):`, or even `if not _is_distributed_initialized():` here?

File: src/weathergen/utils/train_logger.py
Comment (MatKbauer) [inline src/weathergen/utils/train_logger.py:1]: `add_train()` and `add_val()` are fairly similar -> maybe merge into one function with `stage = TRAIN/VAL` argument.

File: src/weathergen/utils/train_logger.py [hunk 3]
```diff
@@ -115,32 +151,38 @@ def add_train(self, samples, lr, loss_avg, stddev_avg, perf_gpu=0.0, perf_mem=0.
...
         self.log_metrics("val", metrics)
         with open(self.path_run / (self.cf.run_id + "_val_log.txt"), "ab") as f:
             np.savetxt(f, log_vals)
 
     #######################################
     @staticmethod
-    def read(run_id, model_path: str, epoch=-1):
+    def read(run_id: str, model_path: str, epoch: int = -1) -> Metrics:
         """
         Read data for run_id
```
Comment (MatKbauer) [inline src/weathergen/utils/train_logger.py:187]: What kind of data?

File: src/weathergen/utils/distributed.py
Comment (clessig) [reply src/weathergen/utils/distributed.py:1]: I have a strong opinion that it's almost always better to keep the context clear.

File: config/streams/streams_ocean/fesom.yml
Comment (kacpnowak) [reply config/streams/streams_ocean/fesom.yml:1]: At one point it was the default dataset for FESOM. Now there's anyway new one waiting to be merged at https://github.com/ecmwf/WeatherGenerator/pull/524

File: src/weathergen/utils/train_logger.py
Comment (clessig) [reply src/weathergen/utils/train_logger.py:1]: This has nothing to do with the PR. Leaving it as is.

File: src/weathergen/utils/train_logger.py (hunk 3 above)
Comment (clessig) [reply src/weathergen/utils/train_logger.py:187]: Unrelated to PR. So leaving as is.

**Comment by clessig** [comment]
@MatKbauer : I went through your comments and did some final adjustments. Could you approve if it's ok for you.

**Review by MatKbauer — APPROVED** [review]
Looks good.
